src/Infrastructure.VS.UnitTests/ThreadHandlingTests.cs

```diff
@@ -113,6 +114,36 @@ public async Task ThrowIfNotOnUIThread_NotUIThread_Throws()
             actual.Should().BeTrue();
         }
 
+        [TestMethod]
+        public async Task RunOnBackgroundThread_StartsOnBackgroundThread_NoSwitch()
+        {
+            await RunOnBackgroundThread(async () =>
+            {
+                var testSubject = new ThreadHandling();
+
+                var startThreadId = System.Threading.Thread.CurrentThread.ManagedThreadId;
+                // Should start on a background thread...
+                VSThreadHelper.CheckAccess().Should().BeFalse();
+
+                var result = await testSubject.RunOnBackgroundThread(() => VSThreadHelper.CheckAccess());
+
+                // Work should be done on the background thread
+                result.Should().BeFalse();
+
+                // ... and finish on a background thread
+                VSThreadHelper.CheckAccess().Should().BeFalse();
+                System.Threading.Thread.CurrentThread.ManagedThreadId.Should().Be(startThreadId);
+            });
+        }
+
+        [TestMethod]
+        [Ignore] // Can't test this case because it requires mocking SwitchToMainThreadAsync which
+                 // is broken. See https://github.com/SonarSource/sonarlint-visualstudio/issues/3144
```

Comment:
We have a helper class `HackedVSThreadingScope` that messes about with VS internals so that calls to `ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync` won't fail.

Unfortunately, it doesn't work in VS2022 (which incidentally means some of the `RunOnUIThread` tests are being ignored. I created #3144 to fix it.

The hack is replacing a field in a VS type using reflection, and the owning type and field name have changed. I know what the new names are, but the RunUIThread tests still fail when you run them all together (they pass individually). At that point I gave up trying to fix it immediately.